include/llvm-dialects/Dialect/OpMap.h

```diff
@@ -0,0 +1,357 @@
+        if (containsCoreOp(op) || containsIntrinsic(op))
+          return true;
+      }
+    }
+
+    for (const DialectOpKV &dialectOpKV : m_dialectOps) {
+      if (dialectOpKV == desc)
+        return true;
+    }
+
+    return false;
+  }
+
+  // -------------------------------------------------------------
+  // get returns a reference to the value stored in any of the
+  // internal map containers or constructs a new value.
+  // -------------------------------------------------------------
+
+  template <typename OpT> ValueT &get() {
+    return get(OpDescription::get<OpT>());
+  }
+
+  // Return a reference to the operation described by a given OpDescription.
+  // This includes any of the possible operation kinds.
+  // Constructs a new entry if the operation was found in neither map.
+  ValueT &get(const OpDescription &desc) {
+    if (desc.isCoreOp() || desc.isIntrinsic()) {
+      assert(desc.getOpcodes().size() == 1 &&
+             "OpMap only supports querying of single core opcodes and "
+             "intrinsics.");
+
+      const unsigned op = desc.getOpcodes().front();
+
+      if (desc.isCoreOp())
+        return m_coreOpcodes[op];
+
+      if (desc.isIntrinsic())
+        return m_intrinsics[op];
+    }
+
+    if (desc.isDialectOp()) {
+      for (DialectOpKV &dialectOpKV : m_dialectOps) {
+        if (dialectOpKV == desc)
+          return dialectOpKV.Value;
+      }
+    }
+
+    // Nothing found. Insert a new dialect operation.
+    return insertDialectOp(desc, {});
+  }
+
+  // -------------------------------------------------------------
+  // Convenience operator[] definition.
+  // -------------------------------------------------------------
+
+  ValueT &operator[](const OpDescription &desc) { return get(desc); }
+
+  // -------------------------------------------------------------
+  // lookup tries to find whether a given function or instruction
+  // can be mapped to any of the entries in the internal map
+  // containers.
+  // It returns either a default-constructed object if the key
+  // was not found or a copy of the contained value.
+  // -------------------------------------------------------------
+
+  // Try to lookup a function which is either the callee of an intrinsic call
+  // or a dialect operation.
+  ValueT lookup(const Function &func) const {
+    if (func.isIntrinsic())
+      return m_intrinsics.lookup(func.getIntrinsicID());
+
+    for (const DialectOpKV &dialectOpKV : m_dialectOps) {
+      const DialectOpKey &key = dialectOpKV.Key;
+      if (detail::isOperationDecl(func.getName(), key.second, key.first))
+        return dialectOpKV.Value;
+    }
+
+    return {};
+  }
+
+  // Try to lookup an instruction which is either an intrinsic instruction,
+  // a dialect operation or a core instruction.
+  ValueT lookup(const Instruction &inst) const {
+    if (auto *CI = dyn_cast<CallInst>(&inst)) {
+      const Function *Callee = CI->getCalledFunction();
+      if (!Callee)
+        return {};
```

Comment:
This case could still fall through to the core opcode case, since `Instruction::Call` could be in `m_coreOpcodes`.